src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -852,6 +852,77 @@ protected theorem extractLsb_ofNat (x n : Nat) (hi lo : Nat) :
     (extractLsb' start len x).getLsbD i = (i < len && x.getLsbD (start+i)) := by
   simp [getLsbD, Nat.lt_succ]
 
+/-
+Get the most significant bit after `extractLsb'`. With `extractLsb'`, we extract
+a `BitVec` `x'` with length `len` from `BitVec` `x`, starting from the element
+at position `start`. The function `getMsb` extracts a bit counting from the most
+significant bit. Assuming certain conditions, `x.(extractLsbD' {w} start len).getMsbD i`
+is equal to `x.getMsbD (w - (start + len - i))`.
+
+Example (w = 9, start := 3, len := 4):
+
+                                |---| = w - (start+len) = 3
+                                      |start + len|     = 7
+                                            |start|     = 4
+                                      |len|             = 3
+let x                       =   9 8 7 6 5 4 3 2 1 0
+let x' = x.extractLsb' 3 4  =       7 6 5 4
+                              |       |
+                              x.getMsbD 0
+                                      | x.getMsbD (i := 4) =
+                                        x'.getMsbD (i := 1) =
+                                        x.getMsbD (i := w - (start + len - i) =
+                                                        10 - (4 + 3 - 1) = 4)
+# Condition 1: `i < len`
+
+The index `i` must be within the range of `len`.
+
+# Condition 2: `start + len - i ≤ w`
+
+If `start + len` is larger than `w`, some lower locations `i` in `x'` are
+outside of `x` such that `getMsbD` returns `false` for these `i`. For
+large enough `i` we will again be within `x`. The precise condition is:
+
+  `start + len - i ≤ w`
+
+w := 10,
+start := 5,                     |    start + len    | = 11
+len := 6 .                                  | start | = 5
+                                |   len   |
+let x                       = _ _ 9 8 7 6 5 4 3 2 1 0
+let x' = x.extractLsb' 5 6  =   _ 9 8 7 6 5 4
+                                |  w - (start + len) + i =
+                                   10 - (    5 +   6) + 0 = -1 -- In Nat, this becomes 0
+                                   start + len - 0 ≤ w
+                                   6 + 5 - 0 = 11 ≤ 10 ❌
```

Comment:
I'm not sure how this calculation relates to `start + len - i ≤ w`. I'm not sure what value of `i` we are currently thinking about (I can infer from what's written that we're thinking of the case where `i = 0`, and also why we are computing `w - (start + len) + i`. I'd have expected that we compute `start + len - i`.

Reply:
I guess there were some inverted indeces in the previous example, that is now fixed and both examples should be consistent.

Reply:
this is hopefully clearer now, too!